test/functional/test_framework/test_framework.py

```diff
@@ -558,19 +558,22 @@ def wait_for_node_exit(self, i, timeout):
         self.nodes[i].process.wait(timeout)
 
     def connect_nodes(self, a, b):
-        from_connection = self.nodes[a]
-        to_connection = self.nodes[b]
-        ip_port = "127.0.0.1:" + str(p2p_port(b))
-        from_connection.addnode(ip_port, "onetry")
-        # poll until version handshake complete to avoid race conditions
-        # with transaction relaying
-        # See comments in net_processing:
-        # * Must have a version message before anything else
-        # * Must have a verack message before anything else
-        wait_until_helper(lambda: all(peer['version'] != 0 for peer in from_connection.getpeerinfo()))
-        wait_until_helper(lambda: all(peer['version'] != 0 for peer in to_connection.getpeerinfo()))
-        wait_until_helper(lambda: all(peer['bytesrecv_per_msg'].pop('verack', 0) == 24 for peer in from_connection.getpeerinfo()))
-        wait_until_helper(lambda: all(peer['bytesrecv_per_msg'].pop('verack', 0) == 24 for peer in to_connection.getpeerinfo()))
+        addr_a = self.nodes[a].p2p_get_bind_address()
```

Comment:
Actually using the bind address to get the destination address is incorrect. The bind address can be `0.0.0.0` in order to allow connection to all network interfaces in the OS (albeit insecure, but the OS won't complain). What you want to do is specifically connect to `127.0.0.1` (or any loopback of your choosing). What you can learn from the bind address is the bind port, but then there should be a function that returns that separately, I'd say.

In fact, this is exactly why I consistenly insisted on calling this a "bind address" everywhere, because this is not really an address... it's a way of telling the OS what network interface(s) you're interested in getting connected through.

Reply:
Also I noticed that you're not using `p2p_port`. Keep in mind that `p2p_port` stores the port that was used to initialize the binding. Meaning, you don't even need the `p2p_get_bind_address` to get the correct port. I think Lukas already has used `p2p_port` to initialize the node? I'm not sure.

Reply:
Localhost is the exception where connecting to `0.0.0.0`/`::1` works because kernel is able to figure out where to send the packets. But I realized that outside localhost that call is not useful, what we're actually interested in is the public-facing address which might be possible to query from libp2p. Anyway, I can change it to connect localhost, I didn't like that because I have to craft the address manually.